woocommerce/woocommerce-android
# Issue/2711 Add downloadable files to a product

Fixes #2711 

**Description**
This PR add the implementation of adding downloadable files flow to products, the list of important changes:

- Adds an entry to the "Add more details" sheet to allow adding downloadable files if the product wasn't downloadable, or didn't have any downloadable files before.
- Created [MediaFilesRepository](https://github.com/woocommerce/woocommerce-android/blob/8ee07126e286d8d8d3c7498bbcd1347462e449f6/WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt) to expose a coroutine for uploading files.
- Modified the viewstates to allow displaying a progress dialog when a file is being uploaded.
- Made some changes to the WPMediaGalleryView to allow single select mode.
- Made some changes to WPMediaPickerFragment to allow defining requestCode in navArgs, and use it for the returned result.
- https://github.com/woocommerce/woocommerce-android/pull/2760/files#r474648485
- Added input validation for the url/name in the file details.

**Screenshots**
<img width="342" alt="Screen Shot 2020-08-21 at 23 14 20" src="https://user-images.githubusercontent.com/1657201/90977586-6eed4d00-e53e-11ea-9a78-678cd6575a7e.png">  ![Screen Shot 2020-08-24 at 11 31 38](https://user-images.githubusercontent.com/1657201/91034795-7cb5d780-e5fd-11ea-9cc3-cb70611a8865.png)
<img width="342" alt="Screen Shot 2020-08-21 at 23 15 05" src="https://user-images.githubusercontent.com/1657201/90977584-6dbc2000-e53e-11ea-949a-51c7eda2a740.png"> <img width="342" alt="Screen Shot 2020-08-21 at 23 15 14" src="https://user-images.githubusercontent.com/1657201/90977583-6dbc2000-e53e-11ea-879e-ea0da9dd56ff.png">
![Screen Shot 2020-08-24 at 11 31 47](https://user-images.githubusercontent.com/1657201/91034855-93f4c500-e5fd-11ea-8780-417cecf37294.png) <img width="342" alt="Screen Shot 2020-08-21 at 23 17 07" src="https://user-images.githubusercontent.com/1657201/90977581-6b59c600-e53e-11ea-876d-2f586c0f411e.png">


**Testing**

**Test 1: Product not downloadable (or downloadable files is empty)**
1. Open product details.
2. Click on "Add more details"
3. Select "Downloadable files"
4. Make sure that a bottom sheet is displayed listing different sources to pick a file.
5. Test different scenarios, and confirm that when picking a file from device or camera, a progress dialog is shown while uploading the file.
6. Edit details of the file in the details screen, and save.
7. Confirm that the file was added.

**Test 2: Product has downloadable files**
1. Open product details.
2. Click on "downloadable files".
3. Click on "Add file"
4. same as steps 4-7 from test 1.
 
**Test 3: Input validation**
Make sure that input validation respects the following cases:
- URL empty => show the error
- URL non empty but doesn’t contain a path (something like: http://testurl.com/):
           - Name is empty => show an error that the name is missing.
           - Name non empty => input is valid, enable the done button.
- URL non empty and have a path (http://testurl.com/file.jpg) => input is valid regardless of the content of the name field.
- URL has to be a valid URL, if the user doesn’t enter a protocol, we’ll add “http” implicitly.
